Clytwynec/get count of workers

clytwynec opened the pull request:
@benpatterson @jzoldak  Please review.

This adds the ability to log the following:
* Queue count for Jenkins instance
* Number of busy executors (according to Jenkins API)
* Number of total executors (according to Jenkins API)
* Number of workers of a particular key_name (according to AWS via boto)

Note that these could be changed to give more detailed info if we want. For example, we could get the number of machines with certain tags from AWS to know how many workers are associated with each master or worker type.

jzoldak commented on aws/get_running_instances.py:
```diff
@@ -0,0 +1,88 @@
+    `key_name` that are running or pending.
+
+    :Args:
+        key_name: the ec2 key_name to match.
```
key_name is a bit tricky, because someone could (intentionally or inadvertently) use that keypair when spinning up a new ami manually to test something out. Though I really don't know what to suggest instead, as the same would hold true for IAM role, etc. Maybe this is just something we need to be aware of. @benpatterson what do you think?

benpatterson replied:
We should be able to configure a read-only user. That's what I think we should do, @jzoldak @clytwynec

clytwynec replied:
A read-only user seems appropriate.

@jzoldak You mean that the data might be misleading because if someone created a worker using that key_name we wouldn't be able to tell that it wasn't created by Jenkins?   I'm not aware of any labels/tags/ids that would distinguish that.

jzoldak replied:
Yeah @clytwynec @benpatterson that's what I meant - if someone spun up a new instance manually using the "jenkins-worker" key then this script wouldn't be able to tell it apart from the ones that jenkins started up. I also can't think of anything else that would be a better distinguisher.

jzoldak commented on aws/get_running_instances.py:
```diff
@@ -0,0 +1,88 @@
+        aws_access_key_id
+        aws_secret_access_key
+    """
+    ec2conn = boto.connect_ec2(aws_access_key_id, aws_secret_access_key)
```
Do we need some defensive programming here, like a try/except block or verifying that ec2conn is not None (or whatever happens when it doesn't get a response from ec2)? I'm not sure what the chance of failure is here.

benpatterson replied:
I also wonder if we let someone use an env var for their aws access key id & secret access key. That is available by convention in both boto and the ec2 API. I think we should support it here.

clytwynec replied:
@benpatterson  The env var is the default for these in the argparse section at the bottom of the file.   So, it tries to get credentials in the following order:
* From the command line opts
* From the environment vars
* From a boto file 

@jzoldak  I didn't add a try except block because it raises boto errors if it fails to connect that describe why it failed, which seem like useful errors to allow through.  (It doesn't return `None`. It returns an Ec2Connection object or raises an exception.)

jzoldak replied:
Awesome. If it raises an exception that's perfect. 
Can you pls add a Raises: section to the docstring?

jzoldak commented on aws/get_running_instances.py:
```diff
@@ -0,0 +1,88 @@
+    parser.add_argument(
+        '--log-level',
+        dest='log_level',
+        help="set logging level",
```
Please add the choices keyword here for extra hardening.

benpatterson commented on aws/get_running_instances.py:
```diff
@@ -0,0 +1,88 @@
+        aws_secret_access_key
+    """
+    ec2conn = boto.connect_ec2(aws_access_key_id, aws_secret_access_key)
+    reservations = ec2conn.get_all_instances()
```
Naming convention: we should use something else rather than 'reservations'. That might confuse it with 'reserved instances' which are a specific kind of EC2 unit (one you buy ahead of time by paying for time in bulk).

clytwynec replied:
Hmm... that's confusing, but `ec2conn.get_all_instances()` does return a list of `Reservation` objects? (Example: `[Reservation:r-abcd1234, Reservation:r-efgh5678]`)

benpatterson replied:
Interesting. It doesn't return "i-abc1234"?

benpatterson commented on aws/get_running_instances.py:
```diff
@@ -0,0 +1,88 @@
+    instances = [
+        i for r in reservations for i in r.instances
+        if i.state in ('running', 'pending') and i.key_name == key_name
+    ]
```
I think we should also have a filter according to tag. We will probably want to use it sooner rather than later.

jzoldak commented on aws/tests.py:
```diff
@@ -0,0 +1,76 @@
+"""
```
Please make a tests folder and move the tests there, to be consistent with the jenkins part of the tree and to be poised for when we inevitable glom more features onto this repo. :)

benpatterson commented on aws/get_running_instances.py:
```diff
@@ -0,0 +1,88 @@
+    reservations = ec2conn.get_all_instances()
+    instances = [
+        i for r in reservations for i in r.instances
+        if i.state in ('running', 'pending') and i.key_name == key_name
```
Should we also include 'shutting-down'? Not sure how Jenkins counts them.

clytwynec replied:
The ec2 plugin only checks for `running` and `pending`, so I did the same here.  Maybe seeing the number of `shutting-down` could be useful too though?

benpatterson replied:
It could be useful, but I don't think it's as important. I mainly wanted to ask because I wasn't sure what Jenkins pays attention to. If shutting-down is ignored by Jenkins, then we should ignore it in this script, too. So...let's not-include it!

benpatterson commented on aws/tests.py:
```diff
@@ -0,0 +1,76 @@
+            self.key_name, self.key_id, self.secret_key
+        )
+
+    @mock_ec2
```
cool decorator

benpatterson commented on jenkins/workers.py:
```diff
@@ -0,0 +1,94 @@
+"""
+Use the jenkins queue and computer APIs to log info about workers.
```
nit:s/log/capture

benpatterson commented on jenkins/workers.py:
```diff
@@ -0,0 +1,94 @@
+    `python jenkins/workers.py -j https://example.jenkins.com`
+
+Example Output:
+    [INFO] Counts of workers and queue length from Jenkins API:
```
Should we have a timestamp on the log as well?

clytwynec replied:
In the Jenkins console log, there is a timestamp on each line already.  I think adding one here might be noisy.

jzoldak replied:
I may be missing something but I think the timestamp isn't necessary/desirable.
The intent is for this to be run via a jenkins job, and to output this info to the console. Then splunk will index the data. If there are timestamps in the output I'm not sure but I think that splunk might parse each line into a separate event, which we don't want.

clytwynec replied:
Does splunk do that for edx-platform-test-subset logs?

jzoldak replied:
Why do you ask, is there a place where we output timestamps to the console? I took a quick look with this splunk query `index=testeng host="ip-10-0-4-206" source="/var/lib/jenkins/jobs/edx-platform-test-subset/*"` but offhand I don't see any.

clytwynec replied:
The timestamps are included in that job on jenkins (https://build.testeng.edx.org/job/edx-platform-test-subset/77858/console), so I was curious how splunk handled those ones.

jzoldak replied:
I could totally be wrong here but I thought that it's the jenkins display that adds the timestamps. The [raw view](https://build.testeng.edx.org/job/edx-platform-test-subset/77858/consoleText) doesn't have them. 
Though I also don't know how the display could add them in. Probably by magic.

benpatterson replied:
I didn't realize/forgot this was a script that would be run in Jenkins. Let's just be sure it's allocated to a worker that will always have an executor available. Timestamps are not crucial so long as splunk gets them accurately. sounds like it might cause more noise than necessary so this is just an open question. Not a must.

clytwynec commented:
Ok.  I think I've addressed all comments about the code itself.  I'm going to work on adding some more detailed info  (e.g. how many instances per worker tag (for aws) or displayName (for jenkins API)),  and perhaps reformatting the output so that it will be more easily interpreted.

jzoldak commented:
:+1:

benpatterson commented:
:+1:
